funnel/utils/markdown/mdit_plugins/abbr.py

```diff
@@ -0,0 +1,173 @@
+    if len(label) == 0 or len(title) == 0:
+        return False
+
+    if 'abbr' not in state.env:
+        state.env['abbr'] = {}
+
+    if label not in state.env['abbr']:
+        state.env['abbr'][label] = title
+
+    state.line = start_line + 1
+    return True
+
+
+def abbr_replace(state: StateInline):
+    block_tokens = state.tokens
+
+    if 'abbr' not in state.env:
+        return
+
+    labels_re_str = '|'.join(
+        [state.md.utils.escapeRE(k) for k in sorted(state.env['abbr'].keys(), key=len)]
+    )
+
+    simple_re = re.compile('(?:' + labels_re_str + ')')
+
+    match_re_str = (
+        '(^|' + PADDING_CHARACTERS + ')'
+        '(' + labels_re_str + ')'
+        '($|' + PADDING_CHARACTERS + ')'
+    )
+
+    match_re = re.compile(match_re_str)
+
+    j, block_tokens_length = 0, len(block_tokens)
+    while j < block_tokens_length:
+        block_token = block_tokens[j]
+        if block_token.type != 'inline':
+            j += 1
+            continue
+        tokens = block_token.children
+
+        i = len(tokens) - 1  # type: ignore[arg-type]
+        while i >= 0:
+            current_token = tokens[i]  # type: ignore[index]
+            if current_token.type != 'text':
+                i -= 1
+                continue
+
+            current_text = current_token.content
+
+            nodes = []
+
+            if simple_re.search(current_text) is None:
+                i -= 1
+                continue
+
+            next_pos = 0
+            for matches in match_re.finditer(current_text):
+                prefix, match = matches.groups()[:2]
+                prefix_indices, suffix_indices = matches.regs[1:4:2]
+
+                if prefix != '':
+                    token = Token('text', '', 0)
+                    token.content = current_text[next_pos : prefix_indices[1]]
+                    nodes.append(token)
+
+                token = Token('abbr_open', 'abbr', 1)
+                token.attrs['title'] = state.env['abbr'][match]
+                nodes.append(token)
+
+                token = Token('text', '', 0)
+                token.content = match
+                nodes.append(token)
+
+                token = Token('abbr_close', 'abbr', -1)
+                nodes.append(token)
+
+                next_pos = suffix_indices[0]
+
+            if len(nodes) == 0:
+                i -= 1
+                continue
+
+            if next_pos < len(current_text):
+                token = Token('text', '', 0)
+                token.content = current_text[next_pos:]
+                nodes.append(token)
+
+            block_token.children = tokens = state.md.utils.arrayReplaceAt(
+                tokens, i, nodes
+            )
+            i -= 1
+
+        j += 1
+
+
+def abbr_plugin(md: MarkdownIt):
+    md.block.ruler.before(
+        'reference', 'abbr_def', abbr_def, {'alt': ['paragraph', 'reference']}
+    )
+    md.core.ruler.after('linkify', 'abbr_replace', abbr_replace)
```

Comment:
What happens if the `reference` and `linkify` rules are removed? Do these get added in the correct place?

Reply:
`reference` & `linkify` are rules specified by the library. They aren't expected to be removed unless the library authors make that change.
In that case, our tests would fail on upgrade.

However, if you meant what would happen if they are disabled, then in that case they would be in the rule chain sequence, but not be run. It should not effect the plugin in anyway, if so.

Reply:
Okay, this is not a problem then.